src/vos/vos_tree.c

```diff
@@ -962,6 +962,34 @@ key_tree_prepare(struct vos_object *obj, daos_epoch_t epoch,
 
 		kbund.kb_epoch	= DAOS_EPOCH_MAX;
 		rbund.rb_iov	= key;
+		if (tclass == VOS_BTR_DKEY) {
+			struct dtx_handle	*dth = vos_dth_get();
+
+			/* Dkey is the determining factor for on which target
+			 * the data will be stored: the akeys and SV/EV tree
+			 * records belong to the same dkey will be stored on
+			 * the same target. But for different dkeys belong to
+			 * the same redundancy group (for the same replicated
+			 * object), the leaders replica for these different
+			 * dkeys may reside on different targets in the same
+			 * redundancy group.
+			 *
+			 * For VOS_ITER_DKEY operation, if some new dkeys which
+			 * DTXs are committable but related leader replicas are
+			 * on different targets, then in spite of list dkeys on
+			 * which replica, the client will get -DER_INPROGRESS
+			 * and has to retry the list_dkey RPC again and again.
+			 * Further more, if there are more dkeys created during
+			 * the retry, related list_dkey RPC may be hung there
+			 * for very long time.
```

Comment:
So enumeration should just ignore the committables and return committed dkeys only, right?

Reply:
No, the committable dkeys cannot be ignored, they needs to be returned for list_dkey RPC. But if the committable dkeys are created after the list_dkey RPC, in theory, they can be ignored. But from the implementation view, we have to distinguish whether the dkey is created before or after the list_dkey RPC, that will complex the logic.